lib/puppet/functions/gitlab_ci_runner/register.rb

```diff
@@ -0,0 +1,24 @@
+require_relative '../../../puppet_x/gitlab/runner.rb'
+
+# A function that registers a Gitlab runner on a Gitlab instance. Be careful, this will be triggered on noop runs as well!
+Puppet::Functions.create_function(:'gitlab_ci_runner::register') do
+  # @param url The url to your Gitlab instance. Please only provide the host part (e.g https://gitlab.com)
+  # @param token Registration token.
+  # @param additional_options A hash with all additional configuration options for that runner
+  # @return [Struct[{ id => Integer[1], token => String[1], }]] Returns a hash with the runner id and authentcation token
+  # @example Using it as a replacement for the Bolt 'register_runner' task
+  #   puppet apply -e "notice(gitlab_ci_runner::register('https://gitlab.com', 'registration-token'))"
+  #
+  dispatch :register do
+    param 'Stdlib::HTTPUrl', :url
+    param 'String[1]', :token
+    optional_param 'Gitlab_ci_runner::Register', :additional_options
+    return_type 'Struct[{ id => Integer[1], token => String[1], }]'
+  end
+
+  def register(url, token, additional_options = {})
+    PuppetX::Gitlab::Runner.register(url, additional_options.merge('token' => token))
```

Comment:
Something a bit like this maybe?? (untested)

```suggestion
    if closure_scope['facts']['clientnoop']
      Puppet.debug 'Skipping registering with gitlab as running in noop'
      'MOCK_TOKEN'
    else
      PuppetX::Gitlab::Runner.register(url, additional_options.merge('token' => token))
    end
```

Reply:
Trying... :)

```diff
--- lib/puppet/functions/gitlab_ci_runner/register.rb
+++ lib/puppet/functions/gitlab_ci_runner/register.rb
@@ -17,7 +17,7 @@ Puppet::Functions.create_function(:'gitlab_ci_runner::register') do
   end

   def register(url, token, additional_options = {})
-    PuppetX::Gitlab::Runner.register(url, additional_options.merge('token' => token))
+    puts closure_scope['facts'].key?('clientnoop') ? closure_scope['facts']['clientnoop'] : 'NOT FOUND'
   rescue Net::HTTPError => e
     raise "Gitlab runner failed to register: #{e.message}"
   end
```

spec test:
```
$ bundle exec rspec spec/functions/register_spec.rb
...
NOT FOUND
...
```

puppet apply:
```
$ puppet apply --noop --modulepath $PWD -e 'notice(gitlab_ci_runner::register("https://foo", "bar"))'
true
...
```

So this seems to work :yay: Any idea how to mock it in the spec test?

Reply:
In `default_module_facts.yml` ??

Reply:
Was easier than I expected. I thought this is some kind of "special" internal fact like trusted facts.

```ruby
  context 'noop does not register runner and returns dummy token' do
    let(:facts) do
      {
        clientnoop: true
      }
    end

    it { is_expected.to run.with_params(url, regtoken, runner_name).and_return('DUMMY-NOOP-TOKEN') }
  end
  ```